feat: add support for work queues

nick-amplify opened the pull request:
Addresses https://github.com/PrefectHQ/terraform-provider-prefect/issues/131. 

I've copied a lot of functionality from the WorkPool implementation in order to support the crud and import of resources as well as datasource for prefect work queues.

nick-amplify committed:
first pass of creating, updating, destroying work queues

nick-amplify committed:
adding work_queue and work_queues datasources and first tests

nick-amplify committed:
add work queues to provider

nick-amplify committed:
adding new auto generated docs

nick-amplify committed:
adding build file to try custom provider

nick-amplify committed:
updating tests using ephemeral workspace

nick-amplify committed:
removing build file

mitchnielsen commented on docs/data-sources/work_queue.md:
```diff
@@ -0,0 +1,37 @@
+- `concurrency_limit` (Number) The concurrency limit applied to this work pool
+- `description` (String) Description of the work pool
+- `id` (String) Work pool ID (UUID)
+- `name` (String) Name of the work pool
```
It looks like there are a few instances where `work pool` should be replaced with `work queue`, could you please patch those up?

mitchnielsen commented on internal/provider/datasources/work_queue_test.go:
```diff
@@ -0,0 +1,122 @@
+				Config: fixtureAccSingleWorkQueue(workspace.Resource, "test-pool", "test-queue"),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr(singleWorkQueueDatasourceName, "name", "test-queue"),
+					resource.TestCheckResourceAttrSet(singleWorkQueueDatasourceName, "id"),
```
`TestCheckResourceAttrSet` is a valid check, but we could go one step further in most cases and use `TestCheckResourceAttr` which lets you specify not only that it's set, but also lets you pass in the expected value. Could you implement that where it makes sense? Thanks!

nick-amplify replied:
Yeah. I'll switch over now. I copied the test from work_pool_test.go which uses `TestCheckResourceAttrSet`, I wasn't sure if there was a specific reason it was used there instead of checking the actual value as well.

mitchnielsen commented on internal/provider/datasources/work_queues.go:
```diff
@@ -0,0 +1,177 @@
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
```
This looks good - could you add the associated tests for this datasource?

mitchnielsen reviewed (changes requested):
Looking good, thanks again for this contribution 🎉 Just a few comments.

nick-amplify committed:
update tests and refs

nick-amplify committed:
update missed pool->queue refs

nick-amplify committed:
switch to TestCheckResourceAttr for known values

mitchnielsen commented on internal/provider/resources/work_queue.go:
```diff
@@ -0,0 +1,361 @@
+		plan.WorkPoolName.ValueString(),
+	)
+	if err != nil {
+		resp.Diagnostics.Append(helpers.CreateClientErrorDiagnostic("Work Pool", err))
```
```suggestion
		resp.Diagnostics.Append(helpers.CreateClientErrorDiagnostic("Work Queue", err))
```

mitchnielsen commented on internal/provider/datasources/work_queue_test.go:
```diff
@@ -0,0 +1,122 @@
+		PreCheck:                 func() { testutils.AccTestPreCheck(t) },
+		Steps: []resource.TestStep{
+			{
+				// Check that we can query a single work pool
```
```suggestion
				// Check that we can query a single work queue
```

mitchnielsen commented on internal/provider/datasources/work_queue_test.go:
```diff
@@ -0,0 +1,122 @@
+				),
+			},
+			{
+				// Check that we can query multiple work pools
```
```suggestion
				// Check that we can query multiple work queue
```

nick-amplify committed:
add back in prefect resource

nick-amplify committed:
Update internal/provider/datasources/work_queue_test.go

Co-authored-by: Mitchell Nielsen <mitchnielsen@users.noreply.github.com>

nick-amplify committed:
update terraform test fixture spaceing

nick-amplify committed:
update terraform test fixture spaceing, more

nick-amplify committed:
another pool->queue ref

nick-amplify committed:
another ref

nick-amplify committed:
another ref

nick-amplify commented:
@mitchnielsen Thank you for all the comments! I have addressed most of them. 
I updated the work_queue and work_queues datasources tests to get them working. I ran them locally by passing in my local prefect provider, so hopefully things will go smoothly on the prefect CI/CD infra. 

Today, I'm going to get the resources tests working, and hopefully we'll be close to done!

mitchnielsen commented:
Awesome, sound good. Thanks again for the contribution - tag me when you're ready for us to take another look 👍🏼

nick-amplify committed:
update import functionality

nick-amplify committed:
updated work queue resource tests

nick-amplify commented:
@mitchnielsen Ready for another look!

mitchnielsen commented on internal/provider/datasources/work_queue.go:
```diff
@@ -0,0 +1,182 @@
+
+// Schema defines the schema for the data source.
+func (d *WorkQueueDataSource) Schema(_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	// Create a copy of the base attributes
+	// and add the account/workspace ID overrides here
+	// as they are not needed in the work_queues (plural) list
```
Nice 👍🏼

mitchnielsen commented on internal/provider/resources/work_queue.go:
```diff
@@ -0,0 +1,367 @@
+	tfModel.WorkPoolName = types.StringValue(queue.WorkPoolName)
+}
+
+func StringPtr(s string) *string {
+	return &s
+}
```
I don't mind having this since it's so small, but you can use `ptr.To` instead of having to define your own here ([example](https://github.com/PrefectHQ/terraform-provider-prefect/blob/90ce565433daccc51e6ba51b7ae3cc393c332e8c/internal/api/automations_test.go#L36)).

mitchnielsen commented:
I ran the tests from my machine (seeing an odd error in CI I'm looking into) and I got this error:

```
    work_queue_test.go:69: Step 1/5 error: Error running apply: exit status 1

        Error: Error creating Work Pool client

          with prefect_work_pool.terraformaccapk7bqn29k,
          on terraform_plugin_test.tf line 19, in resource "prefect_work_pool" "terraformaccapk7bqn29k":
          19: resource "prefect_work_pool" "terraformaccapk7bqn29k" {

        Could not create Work Pool client, due to error: prefect Cloud endpoints
        require an account_id and workspace_id to be set on either the provider or
        the resource. Please report this issue to the provider developers.
```

Will take a look and try to see if this is a local environment problem or something we need to fix in code

nick-amplify committed:
update prefect work pool ephermal workspace id

nick-amplify committed:
use k8s.io/utils/ptr.To instead of custom func

nick-amplify commented:
> I ran the tests from my machine (seeing an odd error in CI I'm looking into) and I got this error:
> 
> ```
>     work_queue_test.go:69: Step 1/5 error: Error running apply: exit status 1
> 
>         Error: Error creating Work Pool client
> 
>           with prefect_work_pool.terraformaccapk7bqn29k,
>           on terraform_plugin_test.tf line 19, in resource "prefect_work_pool" "terraformaccapk7bqn29k":
>           19: resource "prefect_work_pool" "terraformaccapk7bqn29k" {
> 
>         Could not create Work Pool client, due to error: prefect Cloud endpoints
>         require an account_id and workspace_id to be set on either the provider or
>         the resource. Please report this issue to the provider developers.
> ```
> 
> Will take a look and try to see if this is a local environment problem or something we need to fix in code
> 
> EDIT: CI got the same result now: https://github.com/PrefectHQ/terraform-provider-prefect/actions/runs/12757539058/job/35611693185

I forgot to add these refs back into the prefect work pool creation after I used my local setup: 
```terraform
workspace_id = prefect_workspace.test.id
depends_on = [prefect_workspace.test]
```

I just added them back in. Can you give it another try?

nick-amplify committed:
update test ref

nick-amplify committed:
update terraform fixture spacing

mitchnielsen commented:
Sure thing, looks like the same or similar result: https://github.com/PrefectHQ/terraform-provider-prefect/actions/runs/12777563602/job/35620349342

```
    work_queue_test.go:89: Step 1/2 error: Error running apply: exit status 1
        
        Error: Error creating Work Queue client
        
          with prefect_work_queue.test,
          on terraform_plugin_test.tf line 27, in resource "prefect_work_queue" "test":
          27: resource "prefect_work_queue" "test" {
        
        Could not create Work Queue client, due to error: prefect Cloud endpoints
        require an account_id and workspace_id to be set on either the provider or
        the resource. Please report this issue to the provider developers.
```

I think to start, we need this:

```diff
diff --git a/internal/provider/resources/work_queue_test.go b/internal/provider/resources/work_queue_test.go
index c30cbbb..efa80f2 100644
--- a/internal/provider/resources/work_queue_test.go
+++ b/internal/provider/resources/work_queue_test.go
@@ -39,6 +39,7 @@ resource "prefect_work_queue" "%s" {
   work_pool_name = prefect_work_pool.%s.name
   priority = %d
   description = "%s"
+  workspace_id = prefect_workspace.test.id
 }
 
 `, workspace, workPoolName, workPoolName, poolType, paused, baseJobTemplate, workQueueName, workQueueName, workPoolName, priority, description)
```

I made that change locally and tested, and it looks like the URL that's created for the API call was malformed. It wasn't substituting `prefect_work_pool.xyz.name` from the config to the URL. I tried simply changing that value to "test" (since we don't really need a unique value here, as we're in an ephemeral test workspace already), and then it started printing the URL like this:

`/work_pools/%20test/queues/terraformaccxnaj1kmevm`

note the `%20` - I haven't found out yet where that'd be coming from

nick-amplify committed:
add workspace id

mitchnielsen commented:
Aha, finally found it:

```diff
diff --git a/internal/provider/resources/work_queue_test.go b/internal/provider/resources/work_queue_test.go
index efa80f2..31043b3 100644
--- a/internal/provider/resources/work_queue_test.go
+++ b/internal/provider/resources/work_queue_test.go
@@ -302,7 +302,7 @@ func getWorkQueueImportStateID(workQueueResourceName string, workPoolName string
 		}
 		workQueueName := workQueueResource.Primary.Attributes["name"]
 
-		return fmt.Sprintf("%s, %s,%s", workspaceID, workPoolName, workQueueName), nil
+		return fmt.Sprintf("%s,%s,%s", workspaceID, workPoolName, workQueueName), nil
 	}
 }
 ```

nick-amplify committed:
remove space

mitchnielsen commented on internal/provider/datasources/work_queue_test.go:
```diff
@@ -0,0 +1,137 @@
+					resource.TestCheckResourceAttrSet(multipleWorkQueueDatasourceName, "work_queues.1.priority"),
+					resource.TestCheckResourceAttrSet(multipleWorkQueueDatasourceName, "work_queues.1.description"),
+					resource.TestCheckResourceAttr(multipleWorkQueueDatasourceName, "work_queues.2.name", "test-queue-2"),
+					resource.TestCheckResourceAttr(multipleWorkQueueDatasourceName, "work_queues.2.priority", "3"),
```
So I noticed that CI failed on this test, but it passed for me when I triggered it from my laptop. I triggered it a couple more times and got a failure. So I think this one is flaky because the order of the resources is not always the same.

My first thought here is to refactor this test to use a custom check function that accepts a list of work queue objects, and iterates through them to ensure that each object you expect to exist is present. This would get around any expectations around the order of the objects.

That said, I was looking for other examples of this in our codebase and I saw that [automations tests](https://github.com/nick-amplify/terraform-provider-prefect/blob/add-work-queues/internal/provider/resources/automation_test.go) seem to be able to refer to specific items in the list and make assertions there.

@parkedwards - did you run into any ordering problems there? I'll look into this a bit more in the meantime

parkedwards replied:
hmm - the underlying API response _**might**_ have a non-deterministic ordering behavior; it wouldn't be the first time we ran into this.  if this is the case, then i think `.WorkQueue` may not have consistent ordering

https://github.com/PrefectHQ/terraform-provider-prefect/pull/353/files/2212231f90737a8a921a2d0b0fc541c426e9317a#diff-0673513c828ca9f3c2309508c90f8fde82eb4dba1e1aecd1a4c526af0f3a4f9aR171

parkedwards replied:
ok yeah - so I did run into something like this on the Automation resource, and ended up swapping the list's type to a SetAttribute (instead of a ListAttribute). the original issue was that i would get an inconsistent value error, because the API would return randomly sorted lists 

https://github.com/PrefectHQ/terraform-provider-prefect/pull/339/files

i wonder if you might be able to solve for this by swapping to a SetNestedAttribute for `work_queues`

https://developer.hashicorp.com/terraform/plugin/framework/handling-data/attributes/set-nested

parkedwards replied:
update - doing some testing on the `POST /work_pools/<name>/queues/filter` endpoint, i seem to be getting consistent ordering on the queue response...

mitchnielsen replied:
Mm, that's good to know. Maybe TF itself isn't retaining the order based on the attribute schema

parkedwards replied:
let me take a closer look at this today, see what might be going on

ntorba replied:
@mitchnielsen Can you put this through the test suite again this week?

mitchnielsen replied:
For sure, here's the output: https://github.com/PrefectHQ/terraform-provider-prefect/actions/runs/12812637385/job/35724765884?pr=353

I can take a look some time early this week and try to help out

mitchnielsen replied:
Been looking into this today, will follow up if I find something that works.

mitchnielsen replied:
Here's what I have locally. Tests have been passing consistently.

At a high level, this:
- Changes the `work_queues` field from `types.List` to `types.Set` for some ordering consistency (we could arguably 'promise' ordering, but I'm currently leaning against it and hoping folks find specific work queues by name or some other combination of fields)
- Replaces the individual field checks with a function that will loop through expected work queues and make sure they're all present (so order is irrelevant)

<details>
<summary>Click to expand diff</summary>

```diff
diff --git a/internal/provider/datasources/work_queue_test.go b/internal/provider/datasources/work_queue_test.go
index fba155d..2f9af26 100644
--- a/internal/provider/datasources/work_queue_test.go
+++ b/internal/provider/datasources/work_queue_test.go
@@ -1,11 +1,16 @@
 package datasources_test
 
 import (
+	"context"
 	"fmt"
 	"testing"
 
+	"github.com/google/uuid"
 	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	"github.com/hashicorp/terraform-plugin-testing/terraform"
+	"github.com/prefecthq/terraform-provider-prefect/internal/api"
 	"github.com/prefecthq/terraform-provider-prefect/internal/testutils"
+	"k8s.io/utils/ptr"
 )
 
 func fixtureAccSingleWorkQueue(
@@ -74,7 +79,7 @@ resource "prefect_work_queue" "test_queue2" {
 
 data "prefect_work_queues" "test" {
   work_pool_name = prefect_work_pool.test_multi.name
-  workspace_id = prefect_workspace.test.id 
+  workspace_id = prefect_workspace.test.id
   depends_on = [
     prefect_work_pool.test_multi,
 	prefect_work_queue.test_queue1,
@@ -90,7 +95,9 @@ func TestAccDatasource_work_queue(t *testing.T) {
 	singleWorkQueueDatasourceName := "data.prefect_work_queue.test"
 	multipleWorkQueueDatasourceName := "data.prefect_work_queues.test"
 	workspace := testutils.NewEphemeralWorkspace()
-	
+
+	workQueues := []*api.WorkQueue{}
+
 	resource.ParallelTest(t, resource.TestCase{
 		ProtoV6ProviderFactories: testutils.TestAccProtoV6ProviderFactories,
 		PreCheck:                 func() { testutils.AccTestPreCheck(t) },
@@ -112,26 +119,124 @@ func TestAccDatasource_work_queue(t *testing.T) {
 				Config: fixtureAccMultipleWorkQueue(workspace.Resource, "test-pool-multi", "test-queue", "test-queue-2"),
 				Check: resource.ComposeAggregateTestCheckFunc(
 					resource.TestCheckResourceAttr(multipleWorkQueueDatasourceName, "work_queues.#", "3"),
-					resource.TestCheckResourceAttr(multipleWorkQueueDatasourceName, "work_queues.0.name", "test-queue"), // We added this as priority 1, so it will have the highest
-					resource.TestCheckResourceAttrSet(multipleWorkQueueDatasourceName, "work_queues.0.id"),
-					resource.TestCheckResourceAttrSet(multipleWorkQueueDatasourceName, "work_queues.0.created"),
-					resource.TestCheckResourceAttrSet(multipleWorkQueueDatasourceName, "work_queues.0.created"),
-					resource.TestCheckResourceAttrSet(multipleWorkQueueDatasourceName, "work_queues.0.updated"),
-					resource.TestCheckResourceAttr(multipleWorkQueueDatasourceName, "work_queues.0.is_paused", "false"),
-					resource.TestCheckResourceAttr(multipleWorkQueueDatasourceName, "work_queues.0.priority", "1"),
-					resource.TestCheckResourceAttrSet(multipleWorkQueueDatasourceName, "work_queues.0.description"),
-					resource.TestCheckResourceAttr(multipleWorkQueueDatasourceName, "work_queues.1.name", "default"),
-					resource.TestCheckResourceAttr(multipleWorkQueueDatasourceName, "work_queues.1.priority", "2"),
-					resource.TestCheckResourceAttrSet(multipleWorkQueueDatasourceName, "work_queues.1.id"),
-					resource.TestCheckResourceAttrSet(multipleWorkQueueDatasourceName, "work_queues.1.created"),
-					resource.TestCheckResourceAttrSet(multipleWorkQueueDatasourceName, "work_queues.1.updated"),
-					resource.TestCheckResourceAttrSet(multipleWorkQueueDatasourceName, "work_queues.1.is_paused"),
-					resource.TestCheckResourceAttr(multipleWorkQueueDatasourceName, "work_queues.1.priority", "2"),
-					resource.TestCheckResourceAttrSet(multipleWorkQueueDatasourceName, "work_queues.1.description"),
-					resource.TestCheckResourceAttr(multipleWorkQueueDatasourceName, "work_queues.2.name", "test-queue-2"),
-					resource.TestCheckResourceAttr(multipleWorkQueueDatasourceName, "work_queues.2.priority", "3"),
+					testAccCheckworkQueueExists("prefect_work_pool.test_multi", &workQueues),
+					testAccCheckWorkQueueValues(&workQueues, expectedWorkQueues),
 				),
 			},
 		},
 	})
 }
+
+// testAccCheckworkQueueExists is a Custom Check Function that
+// verifies that the API object was created correctly.
+func testAccCheckworkQueueExists(workPoolResourceName string, workQueues *[]*api.WorkQueue) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		workPoolResource, exists := s.RootModule().Resources[workPoolResourceName]
+		if !exists {
+			return fmt.Errorf("Resource not found in state: %s", workPoolResourceName)
+		}
+
+		workPoolName := workPoolResource.Primary.Attributes["name"]
+
+		workspaceResource, exists := s.RootModule().Resources[testutils.WorkspaceResourceName]
+		if !exists {
+			return fmt.Errorf("Resource not found in state: %s", testutils.WorkspaceResourceName)
+		}
+
+		workspaceID, _ := uuid.Parse(workspaceResource.Primary.ID)
+
+		// Initialize the client with the associated workspaceID
+		// NOTE: the accountID is inherited by the one set in the test environment
+		c, err := testutils.NewTestClient()
+		if err != nil {
+			return fmt.Errorf("error creating new test client: %w", err)
+		}
+
+		workQueuesClient, err := c.WorkQueues(uuid.Nil, workspaceID, workPoolName)
+		if err != nil {
+			return fmt.Errorf("error creating new work queues client: %w", err)
+		}
+
+		fetchedWorkQueues, err := workQueuesClient.List(context.Background(), api.WorkQueueFilter{})
+		if err != nil {
+			return fmt.Errorf("error fetching workQueues: %w", err)
+		}
+
+		if len(fetchedWorkQueues) == 0 {
+			return fmt.Errorf("unable to list any work queues for work pool %s", workPoolName)
+		}
+
+		*workQueues = append(*workQueues, fetchedWorkQueues...)
+
+		return nil
+	}
+}
+
+var expectedWorkQueues = []*api.WorkQueue{
+	{
+		Name:        "default",
+		Priority:    ptr.To(int64(2)),
+		Description: ptr.To("The work pool's default queue."),
+		IsPaused:    false,
+	},
+	{
+		Name:        "test-queue",
+		Priority:    ptr.To(int64(1)),
+		Description: ptr.To("my work queue"),
+		IsPaused:    false,
+	},
+	{
+		Name:        "test-queue-2",
+		Description: ptr.To(""),
+		IsPaused:    false,
+		// Intentionally not setting Priority.
+	},
+}
+
+// testAccCheckWorkQueueValues is a Custom Check Function that
+// verifies that the API object matches the expected values.
+func testAccCheckWorkQueueValues(fetched *[]*api.WorkQueue, expected []*api.WorkQueue) resource.TestCheckFunc {
+	return func(_ *terraform.State) error {
+		if len(*fetched) != len(expected) {
+			return fmt.Errorf("Expected work queues to be %d, got %d", len(expected), len(*fetched))
+		}
+
+		for _, expectedWorkQueue := range expected {
+			found := false
+
+			for _, fetchedWorkQueue := range *fetched {
+				if fetchedWorkQueue.Name != expectedWorkQueue.Name {
+					continue
+				}
+
+				// Mark the work queue as 'found', and check each of the other fields to ensure they match.
+				found = true
+				name := expectedWorkQueue.Name
+
+				if *fetchedWorkQueue.Description != *expectedWorkQueue.Description {
+					return fmt.Errorf("Expected work queue '%s' description to be %s, got %s", name, *expectedWorkQueue.Description, *fetchedWorkQueue.Description)
+				}
+
+				if fetchedWorkQueue.IsPaused != expectedWorkQueue.IsPaused {
+					return fmt.Errorf("Expected work queue '%s' is paused to be %t, got %t", name, expectedWorkQueue.IsPaused, fetchedWorkQueue.IsPaused)
+				}
+
+				// Priority is special because if one is not configured, it will get a value based on the priority of the most recently created work queue.
+				// Because of this, let's only check for matching priority when we configure one.
+				if expectedWorkQueue.Priority != nil {
+					if *fetchedWorkQueue.Priority != *expectedWorkQueue.Priority {
+						return fmt.Errorf("Expected work queue '%s' priority to be %d, got %d", name, *expectedWorkQueue.Priority, *fetchedWorkQueue.Priority)
+					}
+				}
+
+				break
+			}
+
+			if !found {
+				return fmt.Errorf("Expected to find work queue '%s' by name", expectedWorkQueue.Name)
+			}
+		}
+
+		return nil
+	}
+}
diff --git a/internal/provider/datasources/work_queues.go b/internal/provider/datasources/work_queues.go
index a594f9d..4c37327 100644
--- a/internal/provider/datasources/work_queues.go
+++ b/internal/provider/datasources/work_queues.go
@@ -2,8 +2,7 @@ package datasources
 
 import (
 	"context"
-	"sort"
-	
+
 	"github.com/hashicorp/terraform-plugin-framework/attr"
 	"github.com/hashicorp/terraform-plugin-framework/datasource"
 	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
@@ -28,7 +27,7 @@ type WorkQueuesSourceModel struct {
 	WorkPoolName types.String          `tfsdk:"work_pool_name"`
 
 	FilterAny  types.List `tfsdk:"filter_any"`
-	WorkQueues types.List `tfsdk:"work_queues"`
+	WorkQueues types.Set  `tfsdk:"work_queues"`
 }
 
 // NewWorkQueuesDataSource returns a new WorkQueuesDataSource.
@@ -83,7 +82,7 @@ Use this data source to search for multiple Work Queues. Defaults to fetching al
 				Optional:    true,
 				Description: "Work queue IDs (UUID) to search for (work queues with any matching UUID are returned)",
 			},
-			"work_queues": schema.ListNestedAttribute{
+			"work_queues": schema.SetNestedAttribute{
 				Computed:    true,
 				Description: "Work queues returned by the server",
 				NestedObject: schema.NestedAttributeObject{
@@ -125,20 +124,6 @@ func (d *WorkQueuesDataSource) Read(ctx context.Context, req datasource.ReadRequ
 		return
 	}
 
-	// Sort the queues by priority
-	sort.SliceStable(queues, func(i, j int) bool {
-		if queues[i].Priority == nil && queues[j].Priority == nil {
-			return false
-		}
-		if queues[i].Priority == nil {
-			return false
-		}
-		if queues[j].Priority == nil {
-			return true
-		}
-		return *queues[i].Priority < *queues[j].Priority
-	})
-
 	attributeTypes := map[string]attr.Type{
 		"id":                customtypes.UUIDType{},
 		"created":           customtypes.TimestampType{},
@@ -177,13 +162,13 @@ func (d *WorkQueuesDataSource) Read(ctx context.Context, req datasource.ReadRequ
 	}
 
 	// Set the final list value to be returned
-	list, diag := types.ListValue(types.ObjectType{AttrTypes: attributeTypes}, queueObjects)
+	set, diag := types.SetValueFrom(ctx, types.ObjectType{AttrTypes: attributeTypes}, queueObjects)
 	resp.Diagnostics.Append(diag...)
 	if resp.Diagnostics.HasError() {
 		return
 	}
 
-	model.WorkQueues = list
+	model.WorkQueues = set
 
 	resp.Diagnostics.Append(resp.State.Set(ctx, &model)...)
 	if resp.Diagnostics.HasError() {
```

</details>

nick-amplify committed:
add sorting of queues by priority

nick-amplify committed:
updating test checks

dacreify commented:
Sorry just a bit of cheerleading. I'm about to create a bunch of queues manually and am really looking forward to this feature! 📣😄

mitchnielsen commented:
I can't push to this branch, so I'm going to close this one in favor of https://github.com/PrefectHQ/terraform-provider-prefect/pull/384. Thanks again for the majority of the work here @nick-amplify 🤝
